sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,600 @@
+faucet" could be implemented as so:
+
+```scheme
+(define-public (claim-from-faucet)
+  (if (isnull? (fetch-entry claimed-before (tuple #sender tx-sender)))
+      (let ((requester tx-sender)) ;; set a local variable requester = tx-sender
+        (insert-entry! claimed-before (tuple #sender requester) (tuple #claimed 'true))
+        (as-contract (stacks-transfer! requester 1)))))
+```
+
+Here, the public function `claim-from-faucet`:
+
+1. Checks if the sender has claimed from the faucet before
+2. Assigns the tx sender to a requester variable
+3. Adds an entry to the tracking map
+4. Uses `as-contract` to send 1 microstack
+
+The primitive function `is-contract?` can be used to determine
+whether a given principal corresponds to a smart contract.
+
+## Stacks Transfer Primitives
+
+To interact with Stacks balances, smart contracts may call the
+`(stacks-transfer!)` function. This function will attempt to transfer
+from the current principal to another principal:
+
+
+```scheme
+(stacks-transfer!
+  to-send-amount
+  recipient-principal)
+```
+
+This function itself _requires_ that the operation have been signed by
+the transfering principal. The `integer` type in our smart contracting
+language is an 8-byte unsigned integer, which allows it to specify the
+maximum amount of microstacks spendable in a single Stacks transfer.
+
+Like any other public smart contract function, this function call
+returns true if the transfer was successful, and false otherwise.
+
+## Data-Space Primitives
+
+Data within a smart contract's data-space is stored within
+`maps`. These stores relate a typed-tuple to another typed-tuple
+(almost like a typed key-value store). As opposed to a table data
+structure, a map will only associate a given key with exactly one
+value. Values in a given mapping are set or fetched using:
+
+1. `(fetch-entry map-name key-tuple)` - This fetches the value
+  associated with a given key in the map, or returns `'null` if there
+  is no such value.
+2. `(set-entry! map-name key-tuple value-tuple)` - This will set the
+  value of `key-tuple` in the data map
+3. `(insert-entry! map-name key-tuple value-tuple)` - This will set
+  the value of `key-tuple` in the data map if and only if an entry
+  does not already exist.
+4. `(delete-entry! map-name key-tuple)` - This will delete `key-tuple`
+   from the data map
+
+We chose to use data maps as opposed to other data structures for two
+reasons:
+
+1. The simplicity of data maps allows for both a simple implementation
+within the VM, and easier reasoning about functions. By inspecting a
+given function definition, it is clear which maps will be modified and
+even within those maps, which keys are affected by a given invocation.
+2. The interface of data maps ensures that the return types of map
+operations are _fixed length_, which is a requirement for static
+analysis of smart contracts' runtime, costs, and other properties.
+
```

Comment:
Just so we're aware of this (and maybe this deserves a comment here) -- these constraints effectively disallow range queries or queries by prefix.  There's no way to iterate over a map within a contract.  This is fine by me, but I think we should set developer expectations on this.

Reply:
Yep, will highlight this.